# Connect to G-Drive

In [1218]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Change Directory in Google Drive

In [1219]:
import os
file_dir = '/content/drive/MyDrive/DS_PROJECT_PORTFOLIO/DS_PROJECT1_DATAPREPROCESSING'
os.chdir(file_dir)

In [1220]:
!ls

'[Complete]Laptop_Price.csv'		    '[Sandbox] DS Project 1 - TUTOR.ipynb'
'[Sandbox] DS Project 1 - MYSANDBOX.ipynb'


# Import Library

In [1221]:
# Dataframe Manipulation
import pandas as pd
import numpy as np

# Data Visualization
import seaborn as sns
import matplotlib.pyplot as plt

# Regex
import regex as re

# Import Dataset

In [1222]:
df = pd.read_csv('[Complete]Laptop_Price.csv')
df.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                Memory                           Gpu  OpSys  Weight  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS  1.37kg   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS  1.34kg   
2            256GB SSD         Intel HD Graphics 620  No OS  1.86kg   
3            512GB SSD            AMD Radeon Pro 455  macOS  1.83kg   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS  1.37kg   

   Price_euros  
0      1339.69  
1       898.94  
2       575.00  
3      2537.45  
4      1803.60

# Data Preprocessing

## Exploratory Data Analysis (EDA)

### Dataset Information

In [1223]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         1303 non-null   int64  
 1   Company           1303 non-null   object 
 2   Product           1303 non-null   object 
 3   TypeName          1303 non-null   object 
 4   Inches            1303 non-null   float64
 5   ScreenResolution  1303 non-null   object 
 6   Cpu               1303 non-null   object 
 7   Ram               1303 non-null   object 
 8   Memory            1303 non-null   object 
 9   Gpu               1303 non-null   object 
 10  OpSys             1303 non-null   object 
 11  Weight            1303 non-null   object 
 12  Price_euros       1303 non-null   float64
dtypes: float64(2), int64(1), object(10)
memory usage: 132.5+ KB


- There's No Null Values
- 1303 Rows and 13 columns
- laptop_ID (int), Inches & Price_euros (float), the rest (object).

In [1224]:
df.describe()

laptop_ID       Inches  Price_euros
count  1303.000000  1303.000000  1303.000000
mean    660.155794    15.017191  1123.686992
std     381.172104     1.426304   699.009043
min       1.000000    10.100000   174.000000
25%     331.500000    14.000000   599.000000
50%     659.000000    15.600000   977.000000
75%     990.500000    15.600000  1487.880000
max    1320.000000    18.400000  6099.000000

- Laptop Inches and Price_euros columns, there's no strange values.
- laptop ID, max values is 1320, in the other hand, this dataset only have 1303 rows

## Feature Engineering (FE)

In [1225]:
df.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                Memory                           Gpu  OpSys  Weight  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS  1.37kg   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS  1.34kg   
2            256GB SSD         Intel HD Graphics 620  No OS  1.86kg   
3            512GB SSD            AMD Radeon Pro 455  macOS  1.83kg   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS  1.37kg   

   Price_euros  
0      1339.69  
1       898.94  
2       575.00  
3      2537.45  
4      1803.60

In [1226]:
df.tail()

laptop_ID Company                              Product  \
1298       1316  Lenovo                       Yoga 500-14ISK   
1299       1317  Lenovo                       Yoga 900-13ISK   
1300       1318  Lenovo                   IdeaPad 100S-14IBR   
1301       1319      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1302       1320    Asus  X553SA-XX031T (N3050/4GB/500GB/W10)   

                TypeName  Inches                            ScreenResolution  \
1298  2 in 1 Convertible    14.0   IPS Panel Full HD / Touchscreen 1920x1080   
1299  2 in 1 Convertible    13.3  IPS Panel Quad HD+ / Touchscreen 3200x1800   
1300            Notebook    14.0                                    1366x768   
1301            Notebook    15.6                                    1366x768   
1302            Notebook    15.6                                    1366x768   

                                       Cpu   Ram              Memory  \
1298            Intel Core i7 6500U 2.5GHz   4GB           128GB SSD   
1299            Intel Core i7 6500U 2.5GHz  16GB           512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz   2GB  64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz   6GB             1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz   4GB           500GB HDD   

                        Gpu       OpSys  Weight  Price_euros  
1298  Intel HD Graphics 520  Windows 10   1.8kg        638.0  
1299  Intel HD Graphics 520  Windows 10   1.3kg       1499.0  
1300      Intel HD Graphics  Windows 10   1.5kg        229.0  
1301     AMD Radeon R5 M330  Windows 10  2.19kg        764.0  
1302      Intel HD Graphics  Windows 10   2.2kg        369.0

To get the insight from dataset, we can extract some insight like:

**ScreenResolution**
- ScreenResolution in ScreenResolution Column
- type of ScreenResolution in ScreenResolution Column
- Touchscreen / Not

**CPU**
- Type of CPU
- CPU Speed

**Ram**
- Ram Size

**Memory**
- Type of Memory
- Memory Size

**Weight**
- Changing the columns type into float

In [1227]:
df.columns

Index(['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches',
       'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight',
       'Price_euros'],
      dtype='object')

In [1228]:
# Using a function to lowering columns name
def lowerColumnsName(dataframe):
    new_columns = [col.lower() for col in dataframe.columns]
    dataframe.columns = new_columns
    return dataframe

In [1229]:
lowerColumnsName(df).columns

Index(['laptop_id', 'company', 'product', 'typename', 'inches',
       'screenresolution', 'cpu', 'ram', 'memory', 'gpu', 'opsys', 'weight',
       'price_euros'],
      dtype='object')

In [1230]:
# Or u can write it like this
df.columns.str.lower()

Index(['laptop_id', 'company', 'product', 'typename', 'inches',
       'screenresolution', 'cpu', 'ram', 'memory', 'gpu', 'opsys', 'weight',
       'price_euros'],
      dtype='object')

Notes:
- the function is using list comprehention
- in the other hand, we can use native syntax in python str.lower()


Btw let's compare the speed

In [1231]:
'''
%%timeit
def lowerColumnsName(dataframe):
    new_columns = [col.lower() for col in dataframe.columns]
    dataframe.columns = new_columns
    return dataframe
'''

'\n%%timeit\ndef lowerColumnsName(dataframe):\n    new_columns = [col.lower() for col in dataframe.columns]\n    dataframe.columns = new_columns\n    return dataframe\n'

In [1232]:
# %timeit df.columns.str.lower()

after looking the speed, so we can make a decision what is better to use.

In [1233]:
df = lowerColumnsName(df)

### ScreenResolution FE

in this FE, we want to extract:
- ScreenResolution in ScreenResolution Column
- type of ScreenResolution in ScreenResolution Column
- Touchscreen / Not

In [1234]:
#df.head()

In [1235]:
df['screentype'] = df['screenresolution'].replace(r'((?:Touchscreen))|(\d+(?:x)\d+)|/', '', regex=True)
df['touchscreen'] = df['screenresolution'].str.extract(r'((?:Touchscreen))')
df['resolution'] = df['screenresolution'].str.extract(r'(\d+(?:x)\d+)')

Handling Screentype

In [1236]:
# Only Check Screentype values
# df['screentype'].value_counts(dropna=False)

In [1237]:
# Only Check touchscreen values
# df['touchscreen'].value_counts(dropna=False)

In [1238]:
# Only Check resolution values
# df['resolution'].value_counts(dropna=False)

In [1239]:
df['screentype'] = df['screentype'].str.strip()
df['screentype'] = df['screentype'].replace(r'  ', '', regex=True)

In [1240]:
# Only Check Screentype
# df['screentype'].value_counts()

In [1241]:
df['screentype'] = df['screentype'].replace(r'^(Full HD)|^(IPS Panel)$|^(?:Quad HD(?:\+))|^(4K Ultra HD)', '', regex=True)
df.loc[df['screentype'] == '' , 'screentype'] = np.nan

In [1242]:
# Only Check Screentype
# df['screentype'].value_counts(dropna=False)

Handling touchscreen

In [1243]:
df['touchscreen'] = df['touchscreen'].apply(lambda x: 1 if x == 'Touchscreen' else 0)

In [1244]:
#df

drop columns screenresolution

In [1245]:
df.drop(columns='screenresolution', inplace=True)

### CPU FE
We want to Extract:
- Type of CPU
- CPU Speed

In [1246]:
df['cpu_speed(GHz)'] = df['cpu'].str.extract(r'(\d+(?:\.\d+)?GHz)')

In [1247]:
df['cpu_speed(GHz)'] = df['cpu_speed(GHz)'].str.replace('GHz', '')

In [1248]:
df['cpu'] = df['cpu'].str.replace(r'(\d+(?:\.\d+)?GHz)', '', regex=True)

In [1249]:
#df

### RAM Size FE

In [1250]:
df.rename(columns={
    'ram':'ram(GB)'
}, inplace=True)
df['ram(GB)'] = df['ram(GB)'].replace(r'(?:GB)', '', regex=True)

### Weight FE

In [1251]:
df.rename(columns={
    'weight':'weight(kg)'
}, inplace=True)
df['weight(kg)'] = df['weight(kg)'].replace(r'(?:kg)', '', regex=True)

### CPU and GPU Brand FE

In [1252]:
df['cpu_brand'] = df['cpu'].str.extract(r'^(\w+)')
df['gpu_brand'] = df['gpu'].str.extract(r'^(\w+)')

In [1253]:
"""
print(df['cpu_brand'].value_counts())
print('================')
print(df['gpu_brand'].value_counts())
"""

"\nprint(df['cpu_brand'].value_counts())\nprint('================')\nprint(df['gpu_brand'].value_counts())\n"

In [1254]:
df.loc[df['cpu_brand'] == 'Samsung',:]

laptop_id  company          product            typename  inches  \
1191       1209  Samsung  Chromebook Plus  2 in 1 Convertible    12.3   

                          cpu ram(GB)              memory                gpu  \
1191  Samsung Cortex A72&A53        4  32GB Flash Storage  ARM Mali T860 MP4   

          opsys weight(kg)  price_euros screentype  touchscreen resolution  \
1191  Chrome OS       1.15        659.0        NaN            1  2400x1600   

     cpu_speed(GHz) cpu_brand gpu_brand  
1191            2.0   Samsung       ARM

In [1255]:
df.drop(index=1191, inplace=True)
df.loc[df['cpu_brand'] == 'Samsung',:]

Empty DataFrame
Columns: [laptop_id, company, product, typename, inches, cpu, ram(GB), memory, gpu, opsys, weight(kg), price_euros, screentype, touchscreen, resolution, cpu_speed(GHz), cpu_brand, gpu_brand]
Index: []

In [1256]:
print(df['cpu_brand'].value_counts())
print('================')
print(df['gpu_brand'].value_counts())

Intel    1240
AMD        62
Name: cpu_brand, dtype: int64
Intel     722
Nvidia    400
AMD       180
Name: gpu_brand, dtype: int64


### Opsys FE

In [1257]:
df.loc[df['opsys'] == 'No OS', 'opsys'] = np.nan

### Memory FE

In [1258]:
df['memory_1'] = df['memory']
df['memory_1'] = df['memory_1'].str.replace('1.0TB','1TB', regex=True)
df['memory_1'] = df['memory_1'].str.replace('1TB','1000GB')
df['memory_1'] = df['memory_1'].str.replace('2TB','2000GB')
df['memory_1'] = df['memory_1'].str.replace('GB','')
df['memory_1'].value_counts()

256 SSD                         412
1000 HDD                        224
500 HDD                         132
512 SSD                         118
128 SSD +  1000 HDD              94
128 SSD                          76
256 SSD +  1000 HDD              73
32 Flash Storage                 37
2000 HDD                         16
64 Flash Storage                 15
512 SSD +  1000 HDD              14
1000 SSD                         14
256 SSD +  2000 HDD              10
1000 Hybrid                       9
256 Flash Storage                 8
16 Flash Storage                  7
32 SSD                            6
180 SSD                           5
128 Flash Storage                 4
512 SSD +  2000 HDD               3
16 SSD                            3
512 Flash Storage                 2
1000 SSD +  1000 HDD              2
256 SSD +  500 HDD                2
128 SSD +  2000 HDD               2
256 SSD +  256 SSD                2
512 SSD +  256 SSD                1
512 SSD +  512 SSD          

In [1259]:
df['memory2'] = df['memory_1'].str.replace(r' ','')
df['memory2'].value_counts()

256SSD                    412
1000HDD                   224
500HDD                    132
512SSD                    118
128SSD+1000HDD             94
128SSD                     76
256SSD+1000HDD             73
32FlashStorage             37
2000HDD                    16
64FlashStorage             15
512SSD+1000HDD             14
1000SSD                    14
256SSD+2000HDD             10
1000Hybrid                  9
256FlashStorage             8
16FlashStorage              7
32SSD                       6
180SSD                      5
128FlashStorage             4
512SSD+2000HDD              3
16SSD                       3
512FlashStorage             2
1000SSD+1000HDD             2
256SSD+500HDD               2
128SSD+2000HDD              2
256SSD+256SSD               2
512SSD+256SSD               1
512SSD+512SSD               1
64FlashStorage+1000HDD      1
1000HDD+1000HDD             1
32HDD                       1
64SSD                       1
128HDD                      1
240SSD    

In [1260]:
df['memory'] = df['memory'].replace(r' ', '', regex=True)

In [1261]:
memory_1 = []
memory_2 = []
for i in df['memory2']:
    if len(re.findall(r'\+', i)) == 1: # DOUBLE DRIVE
        one = re.findall(r'(\w+)', i)
        memory_1.append(one[0])
        memory_2.append(one[1])
    else: # SINGLE DRIVE
        one = re.findall(r'(\w+)',i)
        memory_1.append(one[0])
        memory_2.append('NaN')

In [1262]:
memory_1_gb = []
memory_1_type = []
for i in memory_1:
    memory_1_type.append(re.findall(r'(\D\w+)',i)[0])
    memory_1_gb.append(re.findall(r'(\d+)',i)[0])

In [1263]:
memory_2_gb = []
memory_2_type = []
for i in memory_2:
    if i != 'NaN':
        memory_2_type.append(re.findall(r'(\D\w+)',i)[0])
        memory_2_gb.append(re.findall(r'(\d+)',i)[0])
    else:
        memory_2_type.append('NaN')
        memory_2_gb.append(0)

In [1264]:
df['memory_1_sto(GB)'] = memory_1_gb
df['memory_1_type'] = memory_1_type
df['memory_2_sto(GB)'] = memory_2_gb
df['memory_2_type'] = memory_2_type

In [1265]:
df['memory_1_sto(GB)'] = df['memory_1_sto(GB)'].astype(float)
df['memory_2_sto(GB)'] = df['memory_2_sto(GB)'].astype(float)

### Filling Missing Values

In [1266]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1302 entries, 0 to 1302
Data columns (total 24 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_id         1302 non-null   int64  
 1   company           1302 non-null   object 
 2   product           1302 non-null   object 
 3   typename          1302 non-null   object 
 4   inches            1302 non-null   float64
 5   cpu               1302 non-null   object 
 6   ram(GB)           1302 non-null   object 
 7   memory            1302 non-null   object 
 8   gpu               1302 non-null   object 
 9   opsys             1236 non-null   object 
 10  weight(kg)        1302 non-null   object 
 11  price_euros       1302 non-null   float64
 12  screentype        341 non-null    object 
 13  touchscreen       1302 non-null   int64  
 14  resolution        1302 non-null   object 
 15  cpu_speed(GHz)    1302 non-null   object 
 16  cpu_brand         1302 non-null   object 


#### Screentype

In [1267]:
df.loc[df['opsys'] == 'macOS', 'screentype'] = df.loc[df['opsys'] == 'macOS', 'screentype'].fillna('IPS Panel Retina Display')

In [1268]:
df_group = pd.DataFrame(df.groupby(by=['screentype', 'product'], as_index=False, dropna=False)['laptop_id'].count(), index=None)

In [1269]:
df_group.rename(columns={
    'laptop_id':'product_count'
}, inplace=True)

In [1270]:
#df_group

- too much null values in screentype, drop columns

#### cpu_speed(GHZ)

In [1271]:
#df.loc[df['cpu_speed(GHz)'].isnull(), :]

In [1272]:
df.loc[df['cpu'] == 'Intel Core i3 6006U 2GHz', 'cpu_speed(GHz)'] = df.loc[df['cpu'] == 'Intel Core i3 6006U 2GHz', 'cpu_speed(GHz)'].fillna(2.0)

In [1273]:
#df.loc[df['cpu_speed(GHz)'].isnull(), :]

In [1274]:
df.loc[df['cpu'] == 'AMD FX 9830P 3GHz', 'cpu_speed(GHz)'] = df.loc[df['cpu'] == 'AMD FX 9830P 3GHz', 'cpu_speed(GHz)'].fillna(3.0)

#### Opsys

In [1275]:
df['opsys'].unique()

array(['macOS', nan, 'Windows 10', 'Mac OS X', 'Linux', 'Android',
       'Windows 10 S', 'Chrome OS', 'Windows 7'], dtype=object)

In [1276]:
df['cpu'] = df['cpu'].str.strip()

In [1277]:
df.loc[df['opsys'].isnull(), 'opsys'] = df.loc[df['opsys'].isnull(), 'opsys'].fillna('Windows 10')

### Drop Columns

In [1278]:
df.drop(columns=['memory', 'gpu', 'cpu', 'memory', 'memory_1', 'memory2', 'screentype'], inplace=True)

### Changing Columns Type

In [1280]:
df['ram(GB)'] = df['ram(GB)'].astype(float)
df['weight(kg)'] = df['weight(kg)'].astype(float)
df['price_euros'] = df['price_euros'].astype(float)
df['touchscreen'] = df['touchscreen'].astype(int)
df['cpu_speed(GHz)'] = df['cpu_speed(GHz)'].astype(float)
df['memory_1_sto(GB)'] = df['memory_1_sto(GB)'].astype(float)
df['memory_2_sto(GB)'] = df['memory_2_sto(GB)'].astype(float)

In [1284]:
df.to_csv('[Cleaned]Laptop_Price.csv', index=False)

# Data Visualization

In [1285]:
df1 = pd.read_csv('[Cleaned]Laptop_Price.csv')
df1.head()

laptop_id company      product   typename  inches  ram(GB)       opsys  \
0          1   Apple  MacBook Pro  Ultrabook    13.3      8.0       macOS   
1          2   Apple  Macbook Air  Ultrabook    13.3      8.0       macOS   
2          3      HP       250 G6   Notebook    15.6      8.0  Windows 10   
3          4   Apple  MacBook Pro  Ultrabook    15.4     16.0       macOS   
4          5   Apple  MacBook Pro  Ultrabook    13.3      8.0       macOS   

   weight(kg)  price_euros  touchscreen resolution  cpu_speed(GHz) cpu_brand  \
0        1.37      1339.69            0  2560x1600             2.3     Intel   
1        1.34       898.94            0   1440x900             1.8     Intel   
2        1.86       575.00            0  1920x1080             2.5     Intel   
3        1.83      2537.45            0  2880x1800             2.7     Intel   
4        1.37      1803.60            0  2560x1600             3.1     Intel   

  gpu_brand  memory_1_sto(GB) memory_1_type  memory_2_sto(GB) memory_2_type  
0     Intel             128.0           SSD               0.0           NaN  
1     Intel             128.0  FlashStorage               0.0           NaN  
2     Intel             256.0           SSD               0.0           NaN  
3       AMD             512.0           SSD               0.0           NaN  
4     Intel             256.0           SSD               0.0           NaN